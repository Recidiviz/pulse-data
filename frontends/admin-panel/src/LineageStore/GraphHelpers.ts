// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2025 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { Edge, Node } from "@xyflow/react";

import { BigQueryGraphDisplayNode, GraphReferenceType, NodeUrn } from "./types";
import { throwExpression } from "./Utils";

export function buildAdjacencyMappingFromReferences(
  references: GraphReferenceType[]
): {
  urnToDownstreamUrns: Map<NodeUrn, Set<NodeUrn>>;
  urnToUpstreamUrns: Map<NodeUrn, Set<NodeUrn>>;
} {
  const urnToDownstreamUrns = new Map();
  const urnToUpstreamUrns = new Map();
  references.forEach((r) => {
    // downstream ref map
    if (!urnToDownstreamUrns.has(r.source)) {
      urnToDownstreamUrns.set(r.source, new Set());
    }
    urnToDownstreamUrns.get(r.source)?.add(r.target);

    // upstream ref map
    if (!urnToUpstreamUrns.has(r.target)) {
      urnToUpstreamUrns.set(r.target, new Set());
    }
    urnToUpstreamUrns.get(r.target)?.add(r.source);
  });
  return { urnToDownstreamUrns, urnToUpstreamUrns };
}

function setDifference<T>(setA: Set<T>, setB: Set<T>): Set<T> {
  const result = new Set<T>();
  setA.forEach((item) => {
    if (!setB.has(item)) {
      result.add(item);
    }
  });
  return result;
}

/**
 * Walks the graph as described by |nodes| and |edges| in order to determine which of
 * |nodes| is not reachable (unconnected) due to the removal of |nodesToFilterOut|. The
 * "main" graph is determined by which nodes are connected to |startingUrn|, not which
 * graph might be bigger. The implementation of this function is a basic BFS.
 */
export function findUnreachableNodes(
  startingUrn: NodeUrn,
  allUrns: Set<NodeUrn>,
  urnToUpstreamUrns: Map<NodeUrn, Set<NodeUrn>>,
  urnToDownstreamUrns: Map<NodeUrn, Set<NodeUrn>>,
  urnsToFilterOut: Set<NodeUrn>
): {
  connectedNotFilteredUrns: Set<NodeUrn>;
  unreachableUrns: Set<NodeUrn>;
} {
  // do bfs, starting w/ startingUrn and expanding outward in all directions
  const queue: NodeUrn[] = [startingUrn];
  const visitedNodes: Set<NodeUrn> = new Set([startingUrn]);

  while (queue.length > 0) {
    const currentUrn = queue.shift();
    if (currentUrn === undefined) break;

    // peek downstream
    const downstreamNodes = setDifference(
      urnToDownstreamUrns.get(currentUrn) ?? new Set<string>(),
      urnsToFilterOut
    );
    if (downstreamNodes) {
      downstreamNodes.forEach((n) => {
        if (!visitedNodes.has(n)) {
          visitedNodes.add(n);
          queue.push(n);
        }
      });
    }

    // peek upstream
    const upstreamNodes = setDifference(
      urnToUpstreamUrns.get(currentUrn) ?? new Set<string>(),
      urnsToFilterOut
    );
    if (upstreamNodes) {
      upstreamNodes.forEach((n) => {
        if (!visitedNodes.has(n)) {
          visitedNodes.add(n);
          queue.push(n);
        }
      });
    }
  }

  return {
    connectedNotFilteredUrns: visitedNodes,
    // we want to return all "unreachable" or disconnected nodes that are not already
    // going to be filtered out, so we do that by doing
    // ((allUrns - reachableUrns) - filteredOutUrns)
    unreachableUrns: setDifference(
      setDifference(allUrns, visitedNodes),
      urnsToFilterOut
    ),
  };
}

export function findLargestConnectedGraphOfPossibleUnconnectedGraphs(
  nodes: Node<BigQueryGraphDisplayNode>[],
  edges: Edge[],
  nodesToFilterOut: Node<BigQueryGraphDisplayNode>[]
): Set<NodeUrn> {
  if (nodes.length === 0 || nodes.length === nodesToFilterOut.length) {
    return new Set();
  }
  // first, we build the adj mapping of the entire graph, including the nodes we are
  // filtering out
  const { urnToDownstreamUrns, urnToUpstreamUrns } =
    buildAdjacencyMappingFromReferences(edges);
  const urnsToFilterOut = new Set(nodesToFilterOut.map((n) => n.id));
  let candidateUrns = new Set<NodeUrn>(
    nodes.map((n) => n.id).filter((urn) => !urnsToFilterOut.has(urn))
  );
  let largestConnectedGraph = new Set<NodeUrn>();

  // TODO(#46787) can we be smarter about what node to start with -- if we had a good heuristic
  // for choosing the middle of the graph (maybe like which node has most references?
  // but idk if that is actually faster) in O(1) that could speed up this process
  // considerably
  const connectedGraphStartingPoints: Array<NodeUrn> = [
    candidateUrns.values().next().value ??
      throwExpression("Found no candidate urns to find connected graph"),
  ];

  // then, iteratively try to find different connected portions of the graph to find
  // the largest portion
  while (connectedGraphStartingPoints.length > 0) {
    const connectedGraphStartingPointToTry =
      connectedGraphStartingPoints.shift() ??
      throwExpression("Found no new starting points to find connected graph");

    const { connectedNotFilteredUrns, unreachableUrns } = findUnreachableNodes(
      connectedGraphStartingPointToTry,
      candidateUrns,
      urnToUpstreamUrns,
      urnToDownstreamUrns,
      urnsToFilterOut
    );

    if (connectedNotFilteredUrns.size > largestConnectedGraph.size) {
      largestConnectedGraph = connectedNotFilteredUrns;
    }

    // if the size of the largest graph is bigger than all of the remaining unconnected
    // and unfiltered nodes, we have found the biggest possible connected graph
    if (largestConnectedGraph.size >= unreachableUrns.size) {
      break;
    }

    // if not, let's shrink our search size to the set of remaining nodes and try again
    candidateUrns = unreachableUrns;
    connectedGraphStartingPoints.push(
      candidateUrns.values().next().value ??
        throwExpression("Found no candidate urns to find connected graph")
    );
  }

  return largestConnectedGraph;
}
