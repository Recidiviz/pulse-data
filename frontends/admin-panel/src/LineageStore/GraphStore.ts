// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { Edge, Node } from "@xyflow/react";
import { makeAutoObservable, runInAction } from "mobx";

import { buildNewNodeWithDefaults } from "../components/Lineage/GraphNode/NodeBuilder";
import { DagreEngine } from "../components/Lineage/Layout/DagreEngine";
import { LayoutEngine } from "../components/Lineage/Layout/LayoutEngine";
import { LineageRootStore } from "./LineageRootStore";
import {
  GraphDirection,
  GraphDisplayNode,
  LineageNode,
  NodeUrn,
} from "./types";
import { throwExpression } from "./Utils";

/**
 * Store for managing the graph's visual state and behavior.
 */
export class GraphStore {
  // the node that is searched -- should be reflected in the url of the page
  selectedNode?: NodeUrn;

  // nodes that are passed to ReactFlow to render in the graph
  nodes: Node<GraphDisplayNode>[];

  // edges that are passed to ReactFlow to render in the graph
  edges: Edge[];

  // underlying layout engine / algorithm used to re-position nodes in the viewport
  layoutEngine: LayoutEngine;

  constructor(public rootStore: LineageRootStore) {
    this.selectedNode = undefined;
    this.nodes = [];
    this.edges = [];
    this.layoutEngine = new DagreEngine();

    makeAutoObservable(
      this,
      {
        rootStore: false,
      },
      { autoBind: true }
    );
  }

  /**
   * Resets the graph to be a single node.
   */
  resetGraphToActiveNode = async (urn: NodeUrn) => {
    this.selectedNode = urn;
    this.edges = [];
    this.nodes = [
      buildNewNodeWithDefaults(this.rootStore.lineageStore.nodeForUrn(urn)),
    ];
  };

  /**
   * Expands the graph to display the nodes adjacent to |urn| in |direction|.
   */
  expandGraph = async (urn: NodeUrn, direction: GraphDirection) => {
    this.markNodeAsExpanded(urn, direction);
    const expandedNodeUrns = this.rootStore.lineageStore.expand(
      urn,
      direction,
      this.nodes
    );
    this.recalculateNodePositions(expandedNodeUrns);
  };

  // eslint-disable-next-line class-methods-use-this
  contractGraph = async (urn: NodeUrn, direction: GraphDirection) => {
    // TODO(#46345): implement contraction for mvp
  };

  /**
   * Marks a node as being expanded or not, in order to have the expansion visually
   * rendered on the frontend
   */
  markNodeAsExpanded = (
    urnToMarkExpanded: NodeUrn,
    direction: GraphDirection
  ) => {
    const expandedNodeIndex = this.nodes.findIndex(
      (n) => n.id === urnToMarkExpanded
    );
    if (expandedNodeIndex === -1) {
      throwExpression(`Found no nodes for urn: ${urnToMarkExpanded}`);
    }

    if (direction === GraphDirection.UPSTREAM) {
      this.nodes[expandedNodeIndex].data.isExpandedUpstream = true;
    } else if (direction === GraphDirection.DOWNSTREAM) {
      this.nodes[expandedNodeIndex].data.isExpandedDownstream = true;
    }
  };

  /**
   * Given a new set of nodes |newNodes|, recalculates the nodes positions and updates
   * the graph nodes and edges state.
   */
  recalculateNodePositions = async (newNodes: LineageNode[]) => {
    // build new list of nodes and edges
    const allNodesWithoutPositions = [
      ...this.nodes,
      ...newNodes.map((n) => buildNewNodeWithDefaults(n)),
    ];

    const newEdges = this.rootStore.lineageStore.computeEdgesFromNodes(
      allNodesWithoutPositions.map((n) => n.id)
    );

    // recalculate node positions
    const newNodePositionMap = await this.layoutEngine.layout(
      allNodesWithoutPositions,
      newEdges
    );

    // update state with new positions
    const newNodesWithPositions = allNodesWithoutPositions.map((node) => {
      const newPosition =
        newNodePositionMap.get(node.id) ??
        throwExpression(
          `Unknown node attempted to be added to graph ${node.id}`
        );
      const { position: oldPosition, ...nonPositionContents } = node;
      return {
        ...nonPositionContents,
        position: newPosition,
      };
    });

    runInAction(() => {
      this.nodes = newNodesWithPositions;
      this.edges = newEdges;
    });
  };

  /**
   * Expands the subgraph that exist between two nodes
   */
  expandSubGraph = (sourceUrn: NodeUrn, targetUrn: NodeUrn) => {
    this.rootStore.lineageStore
      .fetchBetween(sourceUrn, targetUrn, this.nodes)
      .then((graphUpdates) => {
        this.recalculateNodePositions(graphUpdates);
      });
  };
}
