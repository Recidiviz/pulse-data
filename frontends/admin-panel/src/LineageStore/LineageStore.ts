// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2025 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { Edge } from "@xyflow/react";
import { makeAutoObservable, runInAction } from "mobx";

import { fetchNodes, fetchNodesBetween } from "../AdminPanelAPI/LineageAPI";
import { Hydratable, HydrationState } from "../InsightsStore/types";
import {
  BigQueryLineageNode,
  EdgeId,
  GraphDirection,
  GraphEdge,
  IsExpandedStatuses,
  NodeUrn,
} from "./types";
import { throwExpression } from "./Utils";

/**
 * Store for introspecting about the entire view graph.
 */
export class LineageStore implements Hydratable {
  // map of urn -> entity metadata
  nodes: Map<NodeUrn, BigQueryLineageNode>;

  // map of edge id (e-{source}-{target}) to edge
  edges: Map<EdgeId, GraphEdge>;

  // map of urn -> set of downstream nodes
  urnToDownstreamUrns: Map<NodeUrn, Set<NodeUrn>>;

  // map of urn -> set of upstream nodes
  urnToUpstreamUrns: Map<NodeUrn, Set<NodeUrn>>;

  hydrationState: HydrationState;

  constructor() {
    this.hydrationState = { status: "needs hydration" };
    this.nodes = new Map<NodeUrn, BigQueryLineageNode>();
    this.urnToDownstreamUrns = new Map<NodeUrn, Set<string>>();
    this.urnToUpstreamUrns = new Map<NodeUrn, Set<string>>();
    this.edges = new Map<EdgeId, Edge>();

    makeAutoObservable(this, undefined, { autoBind: true });
  }

  /**
   * Computes the set of valid edges that connects the provided |urns|.
   */
  computeEdgesFromNodes = (urns: Set<NodeUrn>): GraphEdge[] => {
    const candidateEdges = new Map<string, GraphEdge>();

    urns.forEach((urn: NodeUrn) => {
      // find edges between node and other active upstream nodes
      const upstreamEdgesToAdd = this.urnToUpstreamUrns
        .get(urn)
        ?.intersection(urns);
      upstreamEdgesToAdd?.forEach((upstreamUrn) => {
        const edgeId = `e-${upstreamUrn}-${urn}`;
        if (!candidateEdges.has(edgeId)) {
          candidateEdges.set(
            edgeId,
            this.edges.get(edgeId) ?? throwExpression(`Unknown edge: ${edgeId}`)
          );
        }
      });
      // find edges between node and other active downstream nodes
      const downstreamEdgesToAdd = this.urnToDownstreamUrns
        .get(urn)
        ?.intersection(urns);
      downstreamEdgesToAdd?.forEach((downstreamUrn) => {
        const edgeId = `e-${urn}-${downstreamUrn}`;
        if (!candidateEdges.has(edgeId)) {
          candidateEdges.set(
            edgeId,
            this.edges.get(edgeId) ?? throwExpression(`Unknown edge: ${edgeId}`)
          );
        }
      });
    });

    // sorts the edges so they are returned deterministically so as to avoid unnecessary
    // re-rendering (i think)?
    return Array.from(candidateEdges.values()).sort((a, b) =>
      a.id.localeCompare(b.id)
    );
  };

  computeExpandedUpstreamForNodes = (
    displayedUrns: Set<NodeUrn>,
    hiddenByFiltersUrns: Set<NodeUrn>
  ): Map<NodeUrn, IsExpandedStatuses> => {
    // combine both urns that are displayed and those that are not to determine if the
    // we have expanded all possible adj neighbors
    const allUrnsInDisplayGraph = displayedUrns.union(hiddenByFiltersUrns);
    const isExpandedMap = new Map<string, IsExpandedStatuses>();

    displayedUrns.forEach((urn: NodeUrn) => {
      // if all adj nodes for this node are in the graph, declare it expanded
      isExpandedMap.set(urn, {
        isExpandedDownstream:
          this.urnToDownstreamUrns.get(urn)?.difference(allUrnsInDisplayGraph)
            .size === 0,
        isExpandedUpstream:
          this.urnToUpstreamUrns.get(urn)?.difference(allUrnsInDisplayGraph)
            .size === 0,
      });
    });

    return isExpandedMap;
  };

  /**
   * Retrieves a set of adjacent nodes for a given node URN and direction.
   */
  adjacentNodes = (urn: NodeUrn, direction: GraphDirection): Set<NodeUrn> => {
    switch (direction) {
      case GraphDirection.UPSTREAM:
        return this.urnToUpstreamUrns.get(urn) ?? new Set();
      case GraphDirection.DOWNSTREAM:
        return this.urnToDownstreamUrns.get(urn) ?? new Set();
      default: {
        const exhaustiveCheck: never = direction;
        throwExpression(`Unhandled role: ${exhaustiveCheck}`);
      }
    }
  };

  /**
   * Expands the graph by fetching the adjacent nodes and computing the new edges to add.
   */
  expand = (
    urnToExpand: NodeUrn,
    direction: GraphDirection,
    currentUrns: Set<NodeUrn>
  ): BigQueryLineageNode[] => {
    const adjacentNodes = this.adjacentNodes(urnToExpand, direction);
    return Array.from(adjacentNodes.difference(currentUrns)).map(
      (urn) => this.nodes.get(urn) ?? throwExpression(`Unknown urn: ${urn}`)
    );
  };

  /**
   * Fetches a single node.
   */
  nodeForUrn = (urn: NodeUrn): BigQueryLineageNode => {
    return this.nodes.get(urn) ?? throwExpression(`Unknown urn: ${urn}`);
  };

  /**
   * Fetches all nodes between a start and end node.
   */
  fetchBetweenAndAddToCurrent = async (
    direction: GraphDirection,
    source: NodeUrn,
    target: NodeUrn,
    currentUrns: Set<NodeUrn>
  ): Promise<BigQueryLineageNode[]> => {
    const urnsBetween = await fetchNodesBetween(direction, source, target);
    const newNodeUrns = new Set<NodeUrn>(urnsBetween.urns);

    return Array.from(newNodeUrns.difference(currentUrns)).map(this.nodeForUrn);
  };

  fetchBetween = async (
    direction: GraphDirection,
    source: NodeUrn,
    target: NodeUrn
  ): Promise<BigQueryLineageNode[]> => {
    const urnsBetween = (await fetchNodesBetween(direction, source, target))
      .urns;
    return urnsBetween.map(this.nodeForUrn);
  };

  setHydrationState(hydrationState: HydrationState) {
    this.hydrationState = hydrationState;
  }

  /**
   * Fetches all nodes and references from the admin panel backend and populates the
   * store's internal data structures.
   */
  populateLineageStore = async () => {
    const graphResponse = await fetchNodes();

    const urnToDownstreamUrns = new Map<NodeUrn, Set<string>>();
    const urnToUpstreamUrns = new Map<NodeUrn, Set<string>>();
    const edges = new Map<EdgeId, Edge>();
    const nodes = new Map<NodeUrn, BigQueryLineageNode>();

    // use graph references to build upstream & downstream ref mapping
    graphResponse.references.forEach((r) => {
      // downstream ref map
      if (!urnToDownstreamUrns.has(r.source)) {
        urnToDownstreamUrns.set(r.source, new Set());
      }
      urnToDownstreamUrns.get(r.source)?.add(r.target);

      // upstream ref map
      if (!urnToUpstreamUrns.has(r.target)) {
        urnToUpstreamUrns.set(r.target, new Set());
      }
      urnToUpstreamUrns.get(r.target)?.add(r.source);

      // create map of edge id to edge source & target
      const edgeId = `e-${r.source}-${r.target}`;
      edges.set(edgeId, {
        id: edgeId,
        source: r.source,
        target: r.target,
      });
    });

    // create map of urn to node metadata
    graphResponse.nodes.forEach((n) =>
      nodes.set(n.urn, {
        ...n,
        hasUpstream: !!urnToUpstreamUrns.get(n.urn)?.size,
        hasDownstream: !!urnToDownstreamUrns.get(n.urn)?.size,
      })
    );
    runInAction(() => {
      this.urnToDownstreamUrns = urnToDownstreamUrns;
      this.urnToUpstreamUrns = urnToUpstreamUrns;
      this.edges = edges;
      this.nodes = nodes;
    });
  };

  hydrate = async () => {
    if (
      this.hydrationState.status === "hydrated" ||
      this.hydrationState.status === "loading"
    )
      return;
    try {
      this.setHydrationState({ status: "loading" });
      await this.populateLineageStore();
      this.setHydrationState({ status: "hydrated" });
    } catch (e) {
      this.setHydrationState({ status: "failed", error: e as Error });
    }
  };
}
