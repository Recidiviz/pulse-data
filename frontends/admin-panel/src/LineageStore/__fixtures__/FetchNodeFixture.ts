// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2025 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { GraphType } from "../types";

export const rawGraphInfoFixture: GraphType = {
  nodes: [
    {
      urn: "fake-dataset.fake-table",
    },
    {
      urn: "fake-dataset.fake-table-2",
    },
    {
      urn: "fake-dataset.fake-table-3",
    },
  ],
  references: [
    {
      sourceUrn: "fake-dataset.fake-table",
      targetUrn: "fake-dataset.fake-table-2",
    },
    {
      sourceUrn: "fake-dataset.fake-table",
      targetUrn: "fake-dataset.fake-table-3",
    },
  ],
};
