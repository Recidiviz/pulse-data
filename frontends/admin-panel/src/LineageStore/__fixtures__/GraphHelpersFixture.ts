// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2025 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { Edge, Node } from "@xyflow/react";

import { buildNewBigQueryNodeWithDefaults } from "../../components/Lineage/GraphNode/NodeBuilder";
import { buildAdjacencyMappingFromReferences } from "../GraphHelpers";
import {
  BigQueryGraphDisplayNode,
  BigQueryNodeType,
  GraphType,
  NodeUrn,
} from "../types";

export type TestGraphType = {
  nodes: Node<BigQueryGraphDisplayNode>[];
  edges: Edge[];
  urnToDownstreamUrns: Map<NodeUrn, Set<NodeUrn>>;
  urnToUpstreamUrns: Map<NodeUrn, Set<NodeUrn>>;
};

export function parseGraph(graph: GraphType): TestGraphType {
  return {
    nodes: graph.nodes.map((n) =>
      buildNewBigQueryNodeWithDefaults({
        ...n,
        // for graphHelpers tests so far, these properties are unused
        hasDownstream: true,
        hasUpstream: true,
      })
    ),
    edges: graph.references.map((r) => ({
      id: `e-${r.source}-${r.target}`,
      source: r.source,
      target: r.target,
    })),
    ...buildAdjacencyMappingFromReferences(graph.references),
  };
}

//  1     2
//   \   /
//     3
//   /   \
//  4     5
export const xShapedGraph: GraphType = {
  nodes: [
    {
      urn: "dataset_1.table_1",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_1",
      viewId: "table_1",
      stateCode: null,
    },
    {
      urn: "dataset_2.table_2",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_2",
      viewId: "table_2",
      stateCode: null,
    },
    {
      urn: "dataset_3.table_3",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_3",
      viewId: "table_3",
      stateCode: null,
    },
    {
      urn: "dataset_4.table_4",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_4",
      viewId: "table_4",
      stateCode: null,
    },
    {
      urn: "dataset_5.table_5",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_5",
      viewId: "table_5",
      stateCode: null,
    },
  ],
  references: [
    {
      source: "dataset_1.table_1",
      target: "dataset_3.table_3",
    },
    {
      source: "dataset_2.table_2",
      target: "dataset_3.table_3",
    },
    {
      source: "dataset_3.table_3",
      target: "dataset_4.table_4",
    },
    {
      source: "dataset_3.table_3",
      target: "dataset_5.table_5",
    },
  ],
};

//  1     2
//   \   /
//  |  3  |
//   /   \
//  4     5
//   \   /
//     6
export const diamondShapedGraph: GraphType = {
  nodes: [
    {
      urn: "dataset_1.table_1",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_1",
      viewId: "table_1",
      stateCode: null,
    },
    {
      urn: "dataset_2.table_2",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_2",
      viewId: "table_2",
      stateCode: null,
    },
    {
      urn: "dataset_3.table_3",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_3",
      viewId: "table_3",
      stateCode: null,
    },
    {
      urn: "dataset_4.table_4",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_4",
      viewId: "table_4",
      stateCode: null,
    },
    {
      urn: "dataset_5.table_5",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_5",
      viewId: "table_5",
      stateCode: null,
    },
    {
      urn: "dataset_6.table_6",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_6",
      viewId: "table_6",
      stateCode: null,
    },
  ],
  references: [
    {
      source: "dataset_1.table_1",
      target: "dataset_3.table_3",
    },
    {
      source: "dataset_2.table_2",
      target: "dataset_3.table_3",
    },
    {
      source: "dataset_3.table_3",
      target: "dataset_4.table_4",
    },
    {
      source: "dataset_2.table_2",
      target: "dataset_5.table_5",
    },
    {
      source: "dataset_1.table_1",
      target: "dataset_4.table_4",
    },
    {
      source: "dataset_3.table_3",
      target: "dataset_5.table_5",
    },
    {
      source: "dataset_4.table_4",
      target: "dataset_6.table_6",
    },
    {
      source: "dataset_5.table_5",
      target: "dataset_6.table_6",
    },
  ],
};

//  1      2    4     5
//   \    /      \   /
//     3           6
//        \     /
//          10
//           |
//           7
//         /   \
//        8     9
export const lightlyConnectedSubGraphs: GraphType = {
  nodes: [
    {
      urn: "dataset_1.table_1",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_1",
      viewId: "table_1",
      stateCode: null,
    },
    {
      urn: "dataset_2.table_2",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_2",
      viewId: "table_2",
      stateCode: null,
    },
    {
      urn: "dataset_3.table_3",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_3",
      viewId: "table_3",
      stateCode: null,
    },
    {
      urn: "dataset_4.table_4",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_4",
      viewId: "table_4",
      stateCode: null,
    },
    {
      urn: "dataset_5.table_5",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_5",
      viewId: "table_5",
      stateCode: null,
    },
    {
      urn: "dataset_6.table_6",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_6",
      viewId: "table_6",
      stateCode: null,
    },
    {
      urn: "dataset_7.table_7",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset76",
      viewId: "table_7",
      stateCode: null,
    },
    {
      urn: "dataset_8.table_8",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_8",
      viewId: "table_8",
      stateCode: null,
    },
    {
      urn: "dataset_9.table_9",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_9",
      viewId: "table_9",
      stateCode: null,
    },
    {
      urn: "dataset_10.table_10",
      type: BigQueryNodeType.VIEW,
      datasetId: "dataset_9",
      viewId: "table_9",
      stateCode: null,
    },
  ],
  references: [
    {
      source: "dataset_1.table_1",
      target: "dataset_3.table_3",
    },
    {
      source: "dataset_2.table_2",
      target: "dataset_3.table_3",
    },
    {
      source: "dataset_3.table_3",
      target: "dataset_10.table_10",
    },
    {
      source: "dataset_4.table_4",
      target: "dataset_6.table_6",
    },
    {
      source: "dataset_5.table_5",
      target: "dataset_6.table_6",
    },
    {
      source: "dataset_6.table_6",
      target: "dataset_10.table_10",
    },
    {
      source: "dataset_10.table_10",
      target: "dataset_7.table_7",
    },
    {
      source: "dataset_7.table_7",
      target: "dataset_8.table_8",
    },
    {
      source: "dataset_7.table_7",
      target: "dataset_9.table_9",
    },
  ],
};
