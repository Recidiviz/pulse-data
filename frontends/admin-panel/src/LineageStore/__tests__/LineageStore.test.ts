// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2025 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { configure } from "mobx";

import { fetchNodes } from "../../AdminPanelAPI/LineageAPI";
import { rawGraphInfoFixture } from "../__fixtures__/FetchNodeFixture";
import { LineageStore } from "../LineageStore";

jest.mock("../../AdminPanelAPI/LineageAPI");
const mockFetchNodes = fetchNodes as jest.Mock;

let store: LineageStore;

beforeEach(async () => {
  jest.resetModules();
  configure({ safeDescriptors: false });
  store = new LineageStore();
  mockFetchNodes.mockResolvedValue(rawGraphInfoFixture);
});

afterEach(() => {
  jest.restoreAllMocks();
  configure({ safeDescriptors: true });
});

test("hydrate store", async () => {
  expect(store.hydrationState.status).toEqual("needs hydration");
  await store.hydrate();
  expect(store.hydrationState.status).toEqual("hydrated");
});

test("hydration -- error", async () => {
  const error = new Error("hydrationError");
  mockFetchNodes.mockRejectedValue(error);
  await store.hydrate();
  expect(store.hydrationState).toEqual({
    status: "failed",
    error,
  });
});

test("store state", async () => {
  await store.hydrate();
  expect({
    nodes: store.nodes,
    edges: store.edges,
    urnToDownstreamUrns: store.urnToDownstreamUrns,
    urnToUpstreamUrns: store.urnToUpstreamUrns,
  }).toMatchSnapshot();
});
