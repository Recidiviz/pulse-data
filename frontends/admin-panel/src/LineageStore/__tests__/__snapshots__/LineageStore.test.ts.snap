// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store state 1`] = `
Object {
  "edges": Array [
    Array [
      "e-fake-dataset.fake-table-fake-dataset.fake-table-2",
      Object {
        "id": "e-fake-dataset.fake-table-fake-dataset.fake-table-2",
        "source": "fake-dataset.fake-table",
        "target": "fake-dataset.fake-table-2",
      },
    ],
    Array [
      "e-fake-dataset.fake-table-fake-dataset.fake-table-3",
      Object {
        "id": "e-fake-dataset.fake-table-fake-dataset.fake-table-3",
        "source": "fake-dataset.fake-table",
        "target": "fake-dataset.fake-table-3",
      },
    ],
  ],
  "nodes": Array [
    Array [
      "fake-dataset.fake-table",
      Object {
        "hasDownstream": true,
        "hasUpstream": false,
        "urn": "fake-dataset.fake-table",
      },
    ],
    Array [
      "fake-dataset.fake-table-2",
      Object {
        "hasDownstream": false,
        "hasUpstream": true,
        "urn": "fake-dataset.fake-table-2",
      },
    ],
    Array [
      "fake-dataset.fake-table-3",
      Object {
        "hasDownstream": false,
        "hasUpstream": true,
        "urn": "fake-dataset.fake-table-3",
      },
    ],
  ],
  "urnToDownstreamUrns": Array [
    Array [
      "fake-dataset.fake-table",
      Array [
        "fake-dataset.fake-table-2",
        "fake-dataset.fake-table-3",
      ],
    ],
  ],
  "urnToUpstreamUrns": Array [
    Array [
      "fake-dataset.fake-table-2",
      Array [
        "fake-dataset.fake-table",
      ],
    ],
    Array [
      "fake-dataset.fake-table-3",
      Array [
        "fake-dataset.fake-table",
      ],
    ],
  ],
}
`;
